File: destack-py/destack/language/core/builtin/trait.py
from typing import (
    TYPE_CHECKING,
    ClassVar,
    cast,
    dataclass_transform,
)

from destack.language.registry import (
    TRAIT_CLASS_BY_TYPE,
    TRAIT_TYPE_BY_CLASS,
)
from destack.utils.func import get_superclasses

from .common import EnumType, NodeType, TraitType
from .object import BuiltinObject, _process_object_cls
from .property import (
    _PROPERTY_SPECIFIERS,
    _resolve_trait_type,
)

if TYPE_CHECKING:
    from destack.language import (
        ActionDefinition,
        Entity,
        MethodDefinition,
        PermissionDeclaration,
        PermissionDefinition,
    )

# pyright: reportIncompatibleVariableOverride=false


TRAIT_PREFIXES = ("Is",)

# traits where all matching nodes are ordered together


@dataclass_transform(kw_only_default=True, field_specifiers=_PROPERTY_SPECIFIERS)
def builtin_trait(
    trait_type: TraitType | None,
    *,
    is_extensible: bool = False,
    event_types: tuple[NodeType, ...] = (),
    permissions: tuple["PermissionDeclaration", ...] = (),
):
    """Register a class as an Entity trait."""

    def decorate(cls: type) -> type:
        cls, _ = _process_object_cls(
            cls=cast(type["Trait"], cls),
            object_type=None,
            is_frozen=False,
            is_concrete=False,
            is_struct=False,
            is_node=True,
            is_entity=False,
            is_root_node=False,
            is_abstract=True,
            base_type=None,
            traits=(),
        )
        if trait_type is not None:
            cls.metatype = trait_type
        cls.__is_trait__ = True
        cls.__is_abstract__ = True

        # traits cannot have properties
        real_properties = [
            p for p in cls.__properties__.values() if p.is_wired and p.name != "metatype"
        ]
        assert not real_properties, f"Trait {cls.__name__} has properties: {real_properties}"

        # traits
        traits: list[TraitType] = []
        base_traits: list[TraitType] = []
        for base in cls.__bases__:
            if trait := _resolve_trait_type(base.__name__):
                if trait not in base_traits:
                    base_traits.append(trait)
        for superclass in get_superclasses(cls):
            if superclass is cls:
                continue
            if trait := _resolve_trait_type(superclass.__name__):
                if trait not in traits:
                    traits.append(trait)
        cls.__traits__ = tuple(traits)
        cls.__base_traits__ = tuple(base_traits)
        cast(type["Trait"], cls).__is_extensible__ = is_extensible

        # event types
        cls.__base_event_types__ = tuple(event_types)

        # meta
        if permissions:
            from .definition import PermissionDefinition

            cls.__permissions__ = tuple(
                PermissionDefinition.from_declaration(permission) for permission in permissions
            )

        # register
        if trait_type is not None:
            TRAIT_CLASS_BY_TYPE[trait_type] = cls
            TRAIT_TYPE_BY_CLASS[cast(type["Trait"], cls)] = trait_type

        return cls

    return decorate


@builtin_trait(None)
class Trait(Entity if TYPE_CHECKING else BuiltinObject):
    """
    An Entity trait that ascribes some behavior to an Entity.

    Traits may define logic and access, but no properties.
    """

    # meta
    metatype: ClassVar[TraitType]

    # flags
    __is_node__: ClassVar[bool] = True
    __is_trait__: ClassVar[bool] = True
    """Whether this trait can be extended by custom Traits."""
    __is_extensible__: ClassVar[bool] = False

    # content
    """The permissions defined for this Trait."""
    __permissions__: ClassVar[tuple["PermissionDefinition", ...]] = ()
    """The methods defined for this Trait."""
    __methods__: ClassVar[tuple["MethodDefinition", ...]] = ()
    """The actions defined for this Trait."""
    __actions__: ClassVar[tuple["ActionDefinition", ...]] = ()

    # event
    """The base event types of this Trait (directly)."""
    __base_event_types__: ClassVar[tuple[NodeType, ...]] = ()
    """The event types of this trait (directly and indirectly)."""
    __event_types__: ClassVar[tuple[NodeType, ...]] = ()

    # enum
    """The base enum types of this trait (directly)."""
    __base_enum_types__: ClassVar[tuple[EnumType, ...]] = ()
    """The enum types of this trait (directly and indirectly)."""
    __enum_types__: ClassVar[tuple[EnumType, ...]] = ()


@builtin_trait(TraitType.ORDERED, is_extensible=True)
class IsOrdered(Trait):
    """An Entity that can be ordered."""

    pass


@builtin_trait(TraitType.OWNABLE, is_extensible=True)
class IsOwnable(Trait):
    """An Entity that can be owned by an Actor."""

    pass


@builtin_trait(TraitType.OWNED)
class IsOwned(IsOwnable):
    """An Entity that must be owned by an Actor."""

    pass


@builtin_trait(TraitType.JOINABLE, is_extensible=True)
class IsJoinable(Trait):
    """An Entity that can be joined by Actors."""

    pass


@builtin_trait(TraitType.ACTOR)
class IsActor(Trait):
    """An Entity that can be an Actor (can do something)."""

    pass


@builtin_trait(
    TraitType.REACTABLE,
    is_extensible=True,
    event_types=(NodeType.REACTION_EVENT,),
)
class IsReactable(Trait):
    """An Entity that can be reacted to (with Reactions)."""

    pass


@builtin_trait(
    TraitType.STARABLE,
    is_extensible=True,
    event_types=(NodeType.STAR_EVENT,),
)
class IsStarable(Trait):
    """An Entity that can be starred (with Stars)."""

    pass


@builtin_trait(
    TraitType.FOLLOWABLE,
    is_extensible=True,
    event_types=(NodeType.FOLLOW_EVENT,),
)
class IsFollowable(Trait):
    """An Entity that can be followed (with Follows)."""

    pass


@builtin_trait(TraitType.INTERACTIVE, is_extensible=True)
class IsInteractive(Trait):
    """An Entity that can be interacted with."""

    pass


@builtin_trait(TraitType.DRAGGABLE, is_extensible=True)
class IsDraggable(Trait):
    """An Entity that can be dragged."""

    pass


@builtin_trait(TraitType.SELECTABLE, is_extensible=True)
class IsSelectable(Trait):
    """An Entity that can be selected."""

    pass


@builtin_trait(TraitType.RUNNABLE)
class IsRunnable(Trait):
    """An Entity that can be (directly, with Runs)."""

    pass

File: destack-ts/src/language/core/builtin/trait.ts
import { TraitType } from "@destack/language/core/builtin/common";
import type {
  PropertyDefinition,
  TraitDefinition,
} from "@destack/language/core/builtin/definition";
import { registerTraitClass } from "@destack/language/registry";

/** Internal base class for Trait companion objects.*/
export class TraitClass<N = any, T extends TraitType = TraitType> {
  readonly metatype: T;
  __definition__: TraitDefinition;
  __properties__: Record<string, PropertyDefinition>;
  __propertiesByAlias__: Record<string, PropertyDefinition>;
  __propertiesById__: Record<number, PropertyDefinition>;

  constructor(metatype: any) {
    this.metatype = metatype;
    this.__definition__ = null as any; // set later (in finalize)
    this.__properties__ = {};
    this.__propertiesByAlias__ = {};
    this.__propertiesById__ = {};
  }

  /** Get a PropertyDefinition or CustomProperty by name. */
  property(name: string): PropertyDefinition {
    const prop = this.__properties__[name];
    if (!prop) {
      throw new Error(`Property ${name} not found on ${this.constructor.name}`);
    }
    return prop;
  }
}

/* ==== DESTACK_GENERATED_START:TRAIT:300000 ==== */
/**
 * An Entity that can be owned by an Actor.
 */
export interface IsOwnable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be owned by an Actor.
 */
class IsOwnable$Type extends TraitClass<IsOwnable, TraitType.OWNABLE> {}

export const IsOwnable = new IsOwnable$Type(TraitType.OWNABLE);
registerTraitClass(TraitType.OWNABLE, IsOwnable);
/* ==== DESTACK_GENERATED_END:TRAIT:300000 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:300001 ==== */
/**
 * An Entity that must be owned by an Actor.
 */
export interface IsOwned extends IsOwnable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that must be owned by an Actor.
 */
class IsOwned$Type extends TraitClass<IsOwned, TraitType.OWNED> {}

export const IsOwned = new IsOwned$Type(TraitType.OWNED);
registerTraitClass(TraitType.OWNED, IsOwned);
/* ==== DESTACK_GENERATED_END:TRAIT:300001 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:300003 ==== */
/**
 * An Entity that can be joined by Actors.
 */
export interface IsJoinable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be joined by Actors.
 */
class IsJoinable$Type extends TraitClass<IsJoinable, TraitType.JOINABLE> {}

export const IsJoinable = new IsJoinable$Type(TraitType.JOINABLE);
registerTraitClass(TraitType.JOINABLE, IsJoinable);
/* ==== DESTACK_GENERATED_END:TRAIT:300003 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:300004 ==== */
/**
 * An Entity that can be an Actor (can do something).
 */
export interface IsActor {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be an Actor (can do something).
 */
class IsActor$Type extends TraitClass<IsActor, TraitType.ACTOR> {}

export const IsActor = new IsActor$Type(TraitType.ACTOR);
registerTraitClass(TraitType.ACTOR, IsActor);
/* ==== DESTACK_GENERATED_END:TRAIT:300004 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:1400032 ==== */
/**
 * An Entity that can be reacted to (with Reactions).
 */
export interface IsReactable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be reacted to (with Reactions).
 */
class IsReactable$Type extends TraitClass<IsReactable, TraitType.REACTABLE> {}

export const IsReactable = new IsReactable$Type(TraitType.REACTABLE);
registerTraitClass(TraitType.REACTABLE, IsReactable);
/* ==== DESTACK_GENERATED_END:TRAIT:1400032 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:1400030 ==== */
/**
 * An Entity that can be starred (with Stars).
 */
export interface IsStarable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be starred (with Stars).
 */
class IsStarable$Type extends TraitClass<IsStarable, TraitType.STARABLE> {}

export const IsStarable = new IsStarable$Type(TraitType.STARABLE);
registerTraitClass(TraitType.STARABLE, IsStarable);
/* ==== DESTACK_GENERATED_END:TRAIT:1400030 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:1400034 ==== */
/**
 * An Entity that can be followed (with Follows).
 */
export interface IsFollowable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be followed (with Follows).
 */
class IsFollowable$Type extends TraitClass<IsFollowable, TraitType.FOLLOWABLE> {}

export const IsFollowable = new IsFollowable$Type(TraitType.FOLLOWABLE);
registerTraitClass(TraitType.FOLLOWABLE, IsFollowable);
/* ==== DESTACK_GENERATED_END:TRAIT:1400034 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:700001 ==== */
/**
 * An Entity that can be (directly, with Runs).
 */
export interface IsRunnable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be (directly, with Runs).
 */
class IsRunnable$Type extends TraitClass<IsRunnable, TraitType.RUNNABLE> {}

export const IsRunnable = new IsRunnable$Type(TraitType.RUNNABLE);
registerTraitClass(TraitType.RUNNABLE, IsRunnable);
/* ==== DESTACK_GENERATED_END:TRAIT:700001 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:10000 ==== */
/**
 * An Entity that can be ordered.
 */
export interface IsOrdered {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be ordered.
 */
class IsOrdered$Type extends TraitClass<IsOrdered, TraitType.ORDERED> {}

export const IsOrdered = new IsOrdered$Type(TraitType.ORDERED);
registerTraitClass(TraitType.ORDERED, IsOrdered);
/* ==== DESTACK_GENERATED_END:TRAIT:10000 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:2000000 ==== */
/**
 * An Entity that can be interacted with.
 */
export interface IsInteractive {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be interacted with.
 */
class IsInteractive$Type extends TraitClass<IsInteractive, TraitType.INTERACTIVE> {}

export const IsInteractive = new IsInteractive$Type(TraitType.INTERACTIVE);
registerTraitClass(TraitType.INTERACTIVE, IsInteractive);
/* ==== DESTACK_GENERATED_END:TRAIT:2000000 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:2000001 ==== */
/**
 * An Entity that can be dragged.
 */
export interface IsDraggable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be dragged.
 */
class IsDraggable$Type extends TraitClass<IsDraggable, TraitType.DRAGGABLE> {}

export const IsDraggable = new IsDraggable$Type(TraitType.DRAGGABLE);
registerTraitClass(TraitType.DRAGGABLE, IsDraggable);
/* ==== DESTACK_GENERATED_END:TRAIT:2000001 ==== */

/* ==== DESTACK_GENERATED_START:TRAIT:2000002 ==== */
/**
 * An Entity that can be selected.
 */
export interface IsSelectable {
  /* ==== DESTACK_CUSTOM_START ==== */
  // ...
  /* ==== DESTACK_CUSTOM_END ==== */
}

/**
 * An Entity that can be selected.
 */
class IsSelectable$Type extends TraitClass<IsSelectable, TraitType.SELECTABLE> {}

export const IsSelectable = new IsSelectable$Type(TraitType.SELECTABLE);
registerTraitClass(TraitType.SELECTABLE, IsSelectable);
/* ==== DESTACK_GENERATED_END:TRAIT:2000002 ==== */
